### PR TITLE
[TickTick] Add start and due date

### PR DESCRIPTION
**Pull request by Yakitrak**
## Description
Add startDate and dueDates to the task details. Issue raised by myself here and thought I'd do a quick PR for it: https://github.com/raycast/extensions/issues/9841

Please let me know if there is anything I might have missed.

## Screencast

image

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/ticktick/src/utils/date.ts [hunk 1]
```diff
@@ -33,7 +33,13 @@ export const formatToServerDate = (date: Date | moment.Moment | null | undefined
...
+export const formatPrettyDateTime = (date: string, timezone: string, isFloating: boolean, isAllDay: boolean) => {
   if (!date) return "";
-  return moment(date).format("MMM Do, h:mm a");
+  else if (isAllDay) {
+    return moment(date).format("MMM Do");
+  } else if (isFloating) {
+    return moment(date).format("MMM Do, h:mm a");
+  } else {
+    return moment.tz(date, timezone).format("MMM Do, h:mm a");
+  }
```
Comment (viduycheung) [inline extensions/ticktick/src/utils/date.ts:44]: The `date` property within the task object represents the date and time in UTC format, while the corresponding `timezone` field reflects the timezone of the individual who created the task.

When dealing with all-day events or floating dates — dates without specific times — it is crucial to account for both the original timezone (as determined by the task creator) and the local timezone of the user viewing the task. Conversely, dates that include specific times should be formatted according to the local timezone of the current user.

Comment (viduycheung) [reply extensions/ticktick/src/utils/date.ts:44]: Feel free to conduct your own testing:

- Create an all-day task set in a timezone different from your own. It should appear as occurring on the same day, regardless of your local timezone.
- Create a task with a floating date time, which lacks a specific timezone association. This should be displayed at the same time, irrespective of your local timezone.
- Create a task with a specific time set in a timezone other than your own. It should be displayed with the appropriate time offset relative to your local timezone.

Comment (Yakitrak) [reply extensions/ticktick/src/utils/date.ts:47]: Thank you for the review @viduycheung. I've made some changes and think that should cover it. I did try it out but I don't use timezones or floating time setting in normal use so I might have missed something. If I did miss something, any advice on implementation will be really appreciated.

Comment (viduycheung) [reply extensions/ticktick/src/utils/date.ts:47]: You can take a look at this code
``` typescript
export const convertToLocalTimeZoneDate = (
    dateString: string | undefined,
    {
        isAllDay,
        isFloating,
        timeZone
    }: {
        isAllDay?: boolean;
        isFloating?: boolean;
        timeZone?: string;
    }
) => {
    if (!timeZone) {
        return dateString;
    }

    return convertTimeZoneDate(dateString, {
        isAllDay,
        isFloating,
        fromTimeZone: timeZone
    });
};
  
export const convertTimeZoneDate(
    dateString: string | undefined | null,
    {
        isAllDay,
        isFloating,
        fromTimeZone = localTimeZone,
        toTimeZone = localTimeZone
    }: {
        isAllDay?: boolean;
        isFloating?: boolean;
        fromTimeZone?: string;
        toTimeZone?: string;
    }
) => {
    if (!dateString) {
        return null;
    }
    const date = DateUtils.parseUTCString(dateString);
    if (!date) {
        return null;
    }

    if (isAllDay || isFloating) {
        return DateUtils.formatToServerDate(
            convertToLocalDate(date, fromTimeZone, toTimeZone, !isAllDay)
        );
    }

    return dateString;
}

export const convertToLocalDate = (date, fromTimeZone, toTimeZone, hasTime) => {
    const fromMoment = moment.tz(date, fromTimeZone || localTimeZone);
    if (hasTime) {
        return moment
            .tz(
                [
                    fromMoment.year(),
                    fromMoment.month(),
                    fromMoment.date(),
                    fromMoment.hour(),
                    fromMoment.minute()
                ],
                toTimeZone || localTimeZone
            )
            .toDate();
    } else {
        return moment
            .tz(
                [fromMoment.year(), fromMoment.month(), fromMoment.date()],
                toTimeZone || localTimeZone
            )
            .toDate();
    }
};

const startDate = convertToLocalTimeZoneDate(task.startDate);
// format date with converted startDate with the local timezone
moment(startDate).format("xxx")

```

Comment (Yakitrak) [reply extensions/ticktick/src/utils/date.ts:48]: Thank you for sharing some code but did you see my new changes? Are they working as expected for you? I just did an update. So if my understanding is correct:
- The datetime is always saved as UTC so we parse the string as UTC
- Then if its an all day task, we convert it to original timezone and then display the day only
- Or if its a floating task, we display the datetime without any conversion
- Lastly if its not floating or all day, we convert to the local timezone and display the datetime

The code you've shown is a bit different and more complex than my understanding so maybe there is some other nuance I have not understood?

File: extensions/ticktick/src/utils/date.ts [hunk 2]
```diff
@@ -34,15 +34,16 @@ export const formatToServerDate = (date: Date | moment.Moment | null | undefined
...
-
   if (!date) return "";
+  const dateInUTC = moment.utc(date);
+  if (!dateInUTC.isValid()) return "";
   else if (isAllDay) {
+    const datetimeInOriginalTimezone = moment.tz(dateInUTC, timezone);
     return datetimeInOriginalTimezone.format("MMM Do");
   } else if (isFloating) {
-    return moment(date).format("MMM Do, h:mm a");
+    return moment(dateInUTC).format("MMM Do, h:mm a");
```
Comment (viduycheung) [inline extensions/ticktick/src/utils/date.ts:44]:
```typescript
if (isFloating) {
    const datetimeInOriginalTimezone = moment.tz(dateInUTC, timezone);
    return datetimeInOriginalTimezone.format("MMM Do, h:mm a");
}
```

Comment (viduycheung) [reply extensions/ticktick/src/utils/date.ts:44]: the floating time task displays the same date and time, so it should convert like an all-day task

Comment (Yakitrak) [reply extensions/ticktick/src/utils/date.ts]: Ah thank you, updated now

**Review by viduycheung — APPROVED** [review]

File: extensions/ticktick/src/service/osScript.ts [hunk 3]
```diff
@@ -15,6 +20,8 @@ const taskObject2Task = (object: Record<string, unknown>): Task => {
     items: object.items as Task["items"],
     kind: object.kind as Task["kind"],
     tags: (object.tags || []) as Task["tags"],
+    startDate: startDate as Task["startDate"],
```
Comment (viduycheung) [inline extensions/ticktick/src/service/osScript.ts:23]: format date in render, not store the format date in task object

Comment (Yakitrak) [reply extensions/ticktick/src/service/osScript.ts]: Done, thank you

File: extensions/ticktick/src/components/taskItem.tsx [hunk 4]
```diff
@@ -103,6 +124,25 @@ const TaskItem: React.FC<{
                 icon={{ source: Icon.Dot, tintColor: checkboxColor }}
               />
               <List.Item.Detail.Metadata.Separator />
+              {startDate ? (
+                <>
+                  <List.Item.Detail.Metadata.Label
+                    title="Start Date"
```
Comment (viduycheung) [inline extensions/ticktick/src/components/taskItem.tsx:130]: title="Start"

File: extensions/ticktick/src/components/taskItem.tsx [hunk 5]
```diff
@@ -103,6 +124,25 @@ const TaskItem: React.FC<{
...
+                    text={formatPrettyDateTime(startDate, timeZone, isFloating, isAllDay)}
+                  />
+                  <List.Item.Detail.Metadata.Separator />
+                </>
+              ) : null}
+              {dueDate ? (
+                <>
+                  <List.Item.Detail.Metadata.Label
+                    title="Due Date"
```
Comment (viduycheung) [inline extensions/ticktick/src/components/taskItem.tsx:139]: title="End"

File: extensions/ticktick/src/components/taskItem.tsx [hunk 6]
```diff
@@ -103,6 +124,25 @@ const TaskItem: React.FC<{
...
+                    text={formatPrettyDateTime(dueDate, timeZone, isFloating, isAllDay)}
+                  />
+                  <List.Item.Detail.Metadata.Separator />
+                </>
+              ) : null}
```
Comment (viduycheung) [inline extensions/ticktick/src/components/taskItem.tsx:144]: there some date mode logic not cover
1. when `startDate` is undefined, render nothing
2. when only have `startDate` or `startDate` is same as the `dueDate`, render one label( title="Date" text={ formatPrettyDateTime(startDate, timeZone, isFloating, isAllDay)})
3. else render startDate and dueDate

CleanShot 2024-01-05 at 10  05 44@2x
CleanShot 2024-01-05 at 10  05 50@2x

File: extensions/ticktick/CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,8 @@
 # TickTick Changelog
 
+## [Added] - {PR_MERGE_DATE}
```
[the remark this answers is not in the document]
Comment (Yakitrak) [reply extensions/ticktick/CHANGELOG.md:3]: This automatically gets changed to the date when its merged from the merge action Raycast uses

**Review by viduycheung — APPROVED** [review]
The code now seems to be perfect🔥

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**